Update cartographer page structure (#2057)
viamrobotics/docs: docs/services/slam/cartographer/_index.md

[The Cartographer Project](https://github.com/cartographer-project) contains a C++ library that performs dense SLAM.

To use Cartographer with the Viam [SLAM service](/services/slam/), you can use the [`cartographer`](https://app.viam.com/module/viam/cartographer) [modular resource](/modular-resources/).
See [Modular resources](/modular-resources/#the-viam-registry) for instructions on using a module from the Viam registry on your robot.

The source code for this module is available on the [`viam-cartographer` GitHub repository](https://github.com/viamrobotics/viam-cartographer).

{{% alert title="Info" color="info" %}}

Cartographer supports taking **2D LiDAR** or **3D LiDAR** data and optionally **inertial measurement unit (IMU)** and/or **odometry** data as input.

However, currently, the `cartographer` modular resource only supports taking **2D LiDAR** and optionally **IMU** data as input.
Support for taking **3D LiDAR** and **odometry** data as input may be added in the future.

{{% /alert %}}

> ## Requirements

Author: No, I don't think that makes sense. These requirements are still required to get the data to do mapping in the first place.

The UI as is presents what you refer to as online and offline operations as entirely different flows. While that technically may use the same service

> ## Requirements

Reviewer: Ok, that makes sense that you need to have captured the data at some point. Sounds good.

> ## Requirements

Reviewer: Whoops, I didn't see your second comment. I think either would make sense - having a separate requirements section for live robot and offline, or having this requirement at the top because it technically applies to both.

> ## Requirements

Reviewer: (Unresolving since I had resolved - feel free to resolve)

  {{< alert title="SUPPORT" color="note" >}}

  Currently, the `rplidar` and `cartographer` modules support the Linux platform only.

  {{< /alert >}}

- You must physically connect your RPlidar to your robot before configuring the `rplidar` and `cartographer` modules.

  If you have a Viam Rover and are mounting an RPlidar to your rover, be sure to position the RPlidar so that it **faces forward in the direction of travel**, facing in the same direction as the included webcam.
  For example, if you are using the [RPlidar A1](https://www.slamtec.com/en/Lidar/A1) model, mount it to the Rover so that the **pointed** end of the RPlidar mount housing points in the direction of the front of the Rover.

  If you need a **mount plate** for your RPlidar A1 or A3 model, you can 3D print an adapter plate using the following:

  - [RPlidar A1 adapter STL](https://github.com/viamrobotics/Rover-VR1/blob/master/CAD/RPIidarA1_adapter.STL)
  - [RPlidar A3 adapter STL](https://github.com/viamrobotics/Rover-VR1/blob/master/CAD/RPIidarA3_adapter.STL)

> ## Mapping Modes

Reviewer: I think one thing that's confusing is in this "Mapping Modes" section, we're describing the three modes and then showing you (and linking to) how to run them with a _live_ robot. But you can also create and update maps with previously captured data.

Author: Changed this to be different headers. Let me know if that makes sense (still work in progress)

Reviewer: These make sense, but I still slightly feel like it'd be clearest to nest the "Create a new map", "Update existing map", and "Localize only" headers/sections under something that says it's for using a live robot, since the first two also work in "previously captured data" mode

[The Cartographer Project](https://github.com/cartographer-project) contains a C++ library that performs dense SLAM.

> To use Cartographer with the Viam [SLAM service](/services/slam/), you can use the [`cartographer`](https://app.viam.com/module/viam/cartographer) [modular resource](/extend/modular-resources/).

Reviewer 2: ```suggestion
To use Cartographer with the Viam [SLAM service](/services/slam/), you can use the [`cartographer`](https://app.viam.com/module/viam/cartographer) [modular resource](/modular-resources/).
```

> See [Modular resources](/extend/modular-resources/#the-viam-registry) for instructions on using a module from the Viam registry on your robot.

Reviewer 2: ```suggestion
See [Modular resources](/modular-resources/#the-viam-registry) for instructions on using a module from the Viam registry on your robot.
```

{{% alert title="Info" color="info" %}}

Cartographer supports taking **2D LiDAR** or **3D LiDAR** data and optionally **inertial measurement unit (IMU)** and/or **odometry** data as input.

However, currently, the `cartographer` modular resource only supports taking **2D LiDAR** and optionally **IMU** data as input.
Support for taking **3D LiDAR** and **odometry** data as input may be added in the future.

The source code for this module is available on the [`viam-cartographer` GitHub repository](https://github.com/viamrobotics/viam-cartographer).

{{% /alert %}}

## Requirements

- Your robot must have an RPlidar installed to be able to use the `cartographer` module, such as the [RPlidar A1](https://www.slamtec.com/en/Lidar/A1) or [RPlidar A3](https://www.slamtec.com/en/Lidar/A3).

> - In addition, you must [add the `rplidar` module to your robot](/extend/modular-resources/examples/rplidar/) to support the RPlidar hardware, if you have not done so already.

Reviewer 2: ```suggestion
- In addition, you must [add the `rplidar` module to your robot](/modular-resources/examples/rplidar/) to support the RPlidar hardware, if you have not done so already.
```

{{% alert title="Info" color="info" %}}

See Viam's [Pricing](https://www.viam.com/product/pricing) page to understand the costs associated with running Cartographer in the cloud.

{{% /alert %}}

### Create new map

1. Configure your `cartographer` SLAM service

   After installing your physical RPlidar and adding the `rplidar` module as outlined in the [requirements](#requirements), follow the steps below to add the `cartographer` module to your robot:

   {{< tabs name="Create new map">}}
   {{% tab name="Config Builder" %}}
   Follow the instructions below to set up the `cartographer` module on your robot:

1. Navigate to the **Config** tab of your robot's page in [the Viam app](https://app.viam.com).
1. Click on the **Services** subtab and click **Create service** in the lower-left corner.
1. Select **SLAM**, then select `cartographer`.
   You can also search for "cartographer".
1. Click **Add module**, give your service a name of your choice, then click **Create**.
1. In the resulting `SLAM` service configuration pane, first choose `Create new map` as the **Mapping mode**, then configure the rest of the **Attributes** for that mapping mode:

>    - **Camera**: Select the `name` of the camera component that you created when you [added the `rplidar` module to your robot](/extend/modular-resources/examples/rplidar/).

Reviewer 2: ```suggestion
   - **Camera**: Select the `name` of the camera component that you created when you [added the `rplidar` module to your robot](/modular-resources/examples/rplidar/).
```

   You can see that your cloud slam session is in progress from your **Location** page's **SLAM library** tab.

   ![offline mapping maps computing table](/services/slam/offline-mapping-maps-computing-table.png)

   When you are ready to end the slam session, return to your robot's **Control** tab and click **End session**.
   If you do not click **End session**, the slam session will automatically end after 45 minutes.

   Once the session has ended, the map is saved to your **Location** page's **SLAM library** tab.

   ![offline mapping available maps](/services/slam/offline-mapping-available-maps.png)

   You can click `View map` to view the map in a dynamic pointcloud viewer.

   ![slam library view map](/services/slam/slam-library-view-map.png)

### Update an existing map

To update an existing map with new pointcloud data from a new SLAM session, follow the instructions below.
Updating an existing map uses an instance of the `cartographer` module running in the cloud, and _does not_ overwrite the existing map.

1. Configure your `cartographer` SLAM service

   {{< tabs name="Update existing map">}}
   {{% tab name="Config Builder" %}}

>    - First, follow the configuration steps in [Create a new map](#create-a-new-map).
> 
>    - Then, configure **Select map** and **Map version** with the name and version of the map you would like to update.

Author: ```suggestion
   Configure **Select map** and **Map version** with the name and version of the map you would like to update.
   For the other attributes, review the information in [Create a new map](#create-a-new-map).
```

   ![slam RC card first pointcloud](/services/slam/slam-RC-card-first-pointcloud.png)

   You can see that your cloud slam session is in progress from your **Location** page's **SLAM library** tab.

   ![offline mapping maps computing table](/services/slam/offline-mapping-maps-computing-table.png)

   When you are ready to end the slam session, return to your robot's **Control** tab and click **End session**.
   If you do not click **End session**, the slam session will automatically end after 45 minutes.

   Once the session has ended, the map is saved to your **Location** page's **SLAM library** tab.

   ![offline mapping available maps](/services/slam/offline-mapping-available-maps.png)

   You can click `View map` to view the map in a dynamic pointcloud viewer.

   ![slam library view map](/services/slam/slam-library-view-map.png)

### Update an existing map

To update an existing map with new pointcloud data from a new SLAM session, follow the instructions below.
Updating an existing map uses an instance of the `cartographer` module running in the cloud, and _does not_ overwrite the existing map.

> You must have first [created a new map](#create-a-new-map) to be able to update an existing map.

Author: I guess it's also a bit of a tautology so just removing this sentence.
```suggestion
```

   ![slam RC card first pointcloud](/services/slam/slam-RC-card-first-pointcloud.png)

   You can see that your cloud slam session is in progress from your **Location** page's **SLAM library** tab.

   ![offline mapping maps computing table](/services/slam/offline-mapping-maps-computing-table.png)

   When you are ready to end the slam session, return to your robot's **Control** tab and click **End session**.
   If you do not click **End session**, the slam session will automatically end after 45 minutes.

   Once the session has ended, the map is saved to your **Location** page's **SLAM library** tab.

   ![offline mapping available maps](/services/slam/offline-mapping-available-maps.png)

   You can click `View map` to view the map in a dynamic pointcloud viewer.

   ![slam library view map](/services/slam/slam-library-view-map.png)

### Update an existing map

To update an existing map with new pointcloud data from a new SLAM session, follow the instructions below.
Updating an existing map uses an instance of the `cartographer` module running in the cloud, and _does not_ overwrite the existing map.

> You must have first [created a new map](#create-a-new-map) to be able to update an existing map.

Reviewer: [optional] I would remove the link from "created a new map" because you can have created it either using a live robot or with previously captured data, but we're only linking to the live robot instructions

We could also keep it if we're planning to move offline mapping off this page soon anyway

1. Configure your `cartographer` SLAM service

   {{< tabs name="Update existing map">}}
   {{% tab name="Config Builder" %}}

   - First, follow the configuration steps in [Create a new map](#create-a-new-map).

>    - Then, configure **Select map** and **Map version** with the name and version of the map you would like to update.

Reviewer: [optional] It's a little odd to say "Then, configure 'Select map' and 'Map version'" because in the UI, the 'Select map' and 'Map version' boxes come first:

<img width="782" alt="Screenshot 2023-10-20 at 12 05 24 PM" src="https://github.com/viamrobotics/docs/assets/1517691/8eb48e6a-79da-421e-b486-84d434736de8">

That's why I had phrased it as:

> The configuration is similar to the configuration for [creating a new map](https://docs.viam.com/services/slam/cartographer/#create-new-map), except you additionally configure the following Attribute:

> * "Select map", "Map version": Provide the name and version of the map you would like to update. You can see more details about the available maps from this robot’s Location page under the SLAM library tab.

I think people will understand what to do though, and it's okay to leave it this way as well

     You can see more details about the available maps from your robot's **Location** page under the **SLAM library** tab.

   {{% /tab %}}
   {{% tab name="JSON Example" %}}

>    ```json {class="line-numbers linkable-line-numbers"}

Reviewer: do we want a summary of what is being configured, like we have for "Create a new map"'s JSON Example?

         "type": "slam_map"
       }
     ]
   }
   ```

   {{% /tab %}}
   {{< /tabs >}}

   {{< alert title="Tip" color="tip" >}}
   Be aware that data is not only captured when a slam session is running.
   Data capture and the resulting charges continue, unless you [turn off data capture for your sensors](/services/data/configure-data-capture/).
   {{< /alert >}}

   For more information about the configuration attributes, see [Attributes](#attributes).

2. Start a mapping session

   Navigate to the **Control** tab on your robot's page and click on the drop-down menu matching the `name` of the service you created.
   On the cartographer panel, you can start a mapping session.

   When you start a mapping session:

   - Viam spins up an instance of the `cartographer` module in the cloud to execute the Cartographer algorithm.
   - Viam's [Data Capture](/services/data/configure-data-capture/) continuously monitors and captures your robot's sensor data while the robot is running.
   - Cartographer uses the data captured from when you click **Start session** until you click **End session** to create the new map based on the existing map.

>      Once you click **End session**, the map is uploaded to the cloud and visible on your **Location** page under **SLAM library**.

Reviewer: Same comment as above

   Enter a name for your new map and click **Start session**.

>    Wait for the slam session to finish starting up in the cloud, which **takes about 2 minutes**.

Reviewer: This is incorrect: you don't enter a map name since you're updating a map

Also, this ends a little abruptly without saying that the rest is similar to "Create a new map" mode, so perhaps we could say that here.

{{% alert title="Info" color="info" %}}

Cartographer may take several minutes to find your robot's position on the existing map.
In the meantime, your robot will show up at the map's origin (with the `(x,y)` coordinates `(0,0)`).

{{% /alert %}}

### Localize only

In this mode, the `cartographer` module on your robot executes the Cartographer algorithm itself locally to find its position on a map.

1.  Configure your `cartographer` SLAM service

    {{< tabs name="Localize only">}}
    {{% tab name="Config Builder" %}}

The configuration is similar to the configuration for [updating an existing map](#update-an-existing-map), except instead of configuring a `Data capture rate (Hz)` on the camera and movement sensor, set a `Data polling rate (Hz)` on both.
The `cartographer` module on your robot polls the live LiDAR and IMU directly at these rates, whereas the capture rate is only used when data is being sent to the cloud.

    {{% /tab %}}
    {{% tab name="JSON Example" %}}

>     ```json {class="line-numbers linkable-line-numbers"}

Reviewer: This formatting seems messed up when rendered:

<img width="540" alt="Screenshot 2023-10-20 at 12 09 18 PM" src="https://github.com/viamrobotics/docs/assets/1517691/e9fe060a-d9db-4bee-afe0-1a707e889447">

          "version": "0.3.36"
        }
      ],
      "services": [
        {
          "type": "slam",
          "namespace": "rdk",
          "model": "viam:slam:cartographer",
          "attributes": {
            "config_params": {
              "min_range_meters": "0.2",
              "max_range_meters": "25",
              "mode": "2d"
            },
            "camera": {
              "name": "rplidar",
              "data_frequency_hz": "5"
            },
            "enable_mapping": false,
            "use_cloud_slam": false,
            "existing_map": "${packages.slam_map.test-map-1}/internalState.pbstream"
          },
          "name": "slam"
        }
      ],
      "components": [
        {
          "model": "viam:lidar:rplidar",
          "type": "camera",
          "namespace": "rdk",
          "attributes": {},
          "depends_on": [],
          "name": "rplidar"
        }
      ],
      "packages": [
        {
          "type": "slam_map",
          "name": "test-map-1",
          "version": "1697208847",
          "package": "d1c224e8-483e-4cc7-980f-76b89d8fb507/test-map-1"
        }
      ]
    }
    ```

    {{% /tab %}}
    {{< /tabs >}}

>     {{< alert title="Tip" color="tip" >}}

Reviewer: The formatting of this tip box seems a little weird:

<img width="820" alt="Screenshot 2023-10-20 at 12 10 36 PM" src="https://github.com/viamrobotics/docs/assets/1517691/35c35be4-09bb-40e0-bc45-c88c2502e03d">

}

> ```

Author: I don't have that on hand right now, so will leave for a future update

}
```

{{% /tab %}}
{{< /tabs >}}

1. Start a mapping session

   Navigate to the **Control** tab on your robot's page and click on the drop-down menu matching the `name` of the service you created.
   On the cartographer panel, you can start a mapping session.

   When you start a mapping session:

   - Viam spins up an instance of the `cartographer` module in the cloud to execute the Cartographer algorithm.

>      Because Cartographer's algorithm is CPU-intensive when creating or updating a map, in this mode the `cartographer` module on your robot acts as a stub.

Reviewer: I don't think we need to mention that Cartographer is CPU-intensive when creating or updating a map again here, since we already mentioned it in "Mapping Modes"

   - Viam's [Data Capture](/services/data/configure-data-capture/) continuously monitors and captures your robot's sensor data while the robot is running.

>    - Cartographer uses the data captured from when you click **Start session** until you click **End session** to create the map.

Reviewer: [optional] I thought it would be nice to warn users about this as part of describing the configuration, since they'll have to pay for the data that's being captured even when they're not running a slam session, but if you think it's better here that's okay too

Author: good point! I've added notes near the save config instruction now!

Reviewer: awesome!

>    Once you click **End session**, the map is uploaded to the cloud(?) and visible on your **Location** page under **SLAM library**.

Author: @EshaMaharishi if I understand correctly, the map is being built in the cloud, so I think it is odd to say it is the uploaded to the cloud. Would this be correct:

 ```suggestion
   Once you click **End session**, the map is saved and visible on your **Location** page under **SLAM library**.
```

Reviewer: Good point, yes it is being built in the cloud. Hmm, how about `the map is saved to cloud storage`?

}
```

{{% /tab %}}
{{< /tabs >}}

2. Start a mapping session

   Navigate to the **Control** tab on your robot's page and click on the drop-down menu matching the `name` of the service you created.
   On the cartographer panel, you can start a mapping session.

   When you start a mapping session:

   - Viam spins up an instance of the `cartographer` module in the cloud to execute the Cartographer algorithm.

>      Because Cartographer's algorithm is CPU-intensive when creating or updating a map, in this mode the `cartographer` module on your robot acts as a stub.

Reviewer: Same here, I don't think we need to mention that Cartographer is CPU-intensive when creating or updating a map again here, since we already mentioned it in "Mapping Modes"

   - Viam's [Data Capture](/services/data/configure-data-capture/) continuously monitors and captures your robot's sensor data while the robot is running.
   - Cartographer uses the data captured from when you click **Start session** until you click **End session** to create the new map based on the existing map.

>      Once you click **End session**, the map is uploaded to the cloud and visible on your **Location** page under **SLAM library**.

Reviewer: Same optional comment here as above

   Enter a name for your new map and click **Start session**.
   Wait for the slam session to finish starting up in the cloud, which **takes about 2 minutes**.

{{% alert title="Info" color="info" %}}

Cartographer may take several minutes to find your robot's position on the existing map.
In the meantime, your robot will show up at the map's origin (with the `(x,y)` coordinates `(0,0)`).

> You may not (?) want to move your robot until it has finished localizing correctly.

Author: @EshaMaharishi why would a user not want to do that? What happens if they do, does that lead to errors?

Reviewer: Actually, I'm not sure if it will lead to errors. I had thought the scans might get applied to the wrong part of the map if you start moving, but I just realized that that would be true even if you weren't moving. So, I think we could just remove this line.

> You may not (?) want to move your robot until it has finished localizing correctly.

Author: ```suggestion
```

{{% /alert %}}

### Localize only

In this mode, the `cartographer` module on your robot executes the Cartographer algorithm itself locally.

1. Configure your `cartographer` SLAM service

   {{< tabs name="Localize only">}}
   {{% tab name="Config Builder" %}}

The configuration is similar to the configuration for [updating an existing map](#update-an-existing-map), except instead of configuring a `Data capture rate (Hz)` on the camera and movement sensor, set a `Data polling rate (Hz)` on both.

> The `cartographer` module on your robot polls the live LiDAR and IMU directly at these rates, wheras the capture rate is only used when data is being sent to the cloud.

Reviewer: ```suggestion
The `cartographer` module on your robot polls the live LiDAR and IMU directly at these rates, whereas the capture rate is only used when data is being sent to the cloud.
```

}
```

{{% /tab %}}
{{< /tabs >}}

> <!-- prettier-ignore -->

Reviewer: The attributes table seems a little out of place here (in the middle of the "Localize only" section) - should we move it after "Localize only" and give it a header?

Author: let me try putting them with the configuration sections

| `config_params` |  obj | Optional | Parameters available to fine-tune the `cartographer` algorithm: [read more below](#config_params). |

2. Start a mapping session

   Navigate to the **Control** tab on your robot's page and click on the drop-down menu matching the `name` of the service you created.

   Unlike when creating or updating a map, you do not need to start and end a slam session.
   The pointcloud for the existing map will appear **immediately** and Cartographer will try to find your robot's position on it.

   Since the map will not change, nothing new will be added to this robot's location's **SLAM library**.

   ![slam RC card localize only](/services/slam/slam-RC-card-localize-only.png)

   {{% alert title="Info" color="info" %}}

   Cartographer may take several minutes to find your robot's position on the existing map.
   In the meantime, your robot will show up at the map's origin (with the `(x,y)` coordinates `(0,0)`).

   If you move your robot, it will appear to be moving in a trajectory from the map's origin.

   {{% /alert %}}

> ## Create a new map from cloud data

Reviewer: [optional] "Create or update a map from previously captured data"? (mention both create and update, and say "previously captured" rather than "cloud")

The reason I'm suggesting "previously captured" instead of "cloud" is that even when using a live robot, the lidar/imu data gets sent to the cloud

Author: The UI should say that too then, right?

Reviewer: Ahh, I see now. If you click "Make new map" the UI says "Create a new map from cloud data", and if you click "Update map" it says "Update <map name> from cloud data" (for example "Update whole-office-3 from cloud data").

How about "Create or update a map from cloud data"?

Or, feel free to leave it as you see fit and resolve

Author: Changing it to your first suggestion - we should probably just update the UI when we can as well

Reviewer: Sounds good, filed a ticket to change the UI: https://viam.atlassian.net/browse/RSDK-5423

Feel free to resolve!

You can specify a range of **previously captured** LiDAR and optional IMU data to create a map or update an existing map in the cloud.
You can browse your previously captured data from the **Data** page under the **Point clouds** tab (for LiDAR data) and **Sensors** tab (for IMU data).

### Requirements

To create a map, you must have already captured LiDAR data in the location in which you would like to create the map.

The following example shows the previously-captured LiDAR data under the **Point clouds** tab for a robot named `test`.

> Selecting a row opens a pane to the right that contains more information, such as the `Robot ID` of the robot the component belongs to:

Reviewer: Nice addition!

Also, one of the UI improvements I asked Data/ML for was to show the robot name and ID for the "Sensors" tab, which is why the IMU screenshot doesn't show those: https://viam.atlassian.net/browse/DATA-2066
